---
title: "Mouse Neurovascular scRNAseq: Cluster Interrogation"
author: "Joshua Samuels"
date: "12/13/2021"
output: html_document
---

# Below are examples of how the functions are used. Any gene can be tested using the following methods.

### Load Libraries

```{r message=FALSE, warning=FALSE}
# Load libraries
library(SingleCellExperiment)
library(Seurat)
library(tidyverse)
library(Matrix)
library(scales)
library(cowplot)
library(RCurl)
library(BiocManager)
library(ggplot2)
library(dplyr)
library(magrittr)
library(patchwork)
library(ggraph)
library(clustree)
```

## Load Data
```{r}
# Load data
seurat_subset_labeled_full <-readRDS(
          "D:/Research/NIH/Herkenham_Lab/10X_Neurovascular_scRNAseq/Data_Objects/Seurat4.0_subset_labeled_full_condensed_Josh_1.22.rds")

seurat_EC_subset <-readRDS(
          "D:/Research/NIH/Herkenham_Lab/10X_Neurovascular_scRNAseq/Data_Objects/Seurat4.0_EC_subset_labeled_Josh_8.21.rds")

seurat_MG_subset <-readRDS(
          "D:/Research/NIH/Herkenham_Lab/10X_Neurovascular_scRNAseq/Data_Objects/Seurat4.0_MG_subset_labeled_Josh_8.21.rds")

# seurat_ChPlex_subset <-readRDS(
#           "D:/Research/NIH/Herkenham_Lab/10X_Neurovascular_scRNAseq/Data_Objects/Seurat4.0_ChPlex_subset_labeled_Josh_8.21.rds")
```


## Visualize Signature Markers & Gene Expression

```{r}
# Create stacked violin plot
# Used code from this place https://divingintogeneticsandgenomics.rbind.io/post/stacked-violin-plot-for-visualizing-single-cell-data-in-seurat/
# remove the x-axis text and tick
# plot.margin to adjust the white space between each plot.
# ... pass any arguments to VlnPlot in Seurat
modify_vlnplot<- function(obj, 
                          feature, 
                          pt.size = 0, 
                          plot.margin = unit(c(-0.75, 0, -0.75, 0), "cm"),
                          ...) {
  p<- VlnPlot(obj, features = feature, pt.size = pt.size, ... )  + 
    xlab("") + ylab(feature) + ggtitle("") + 
    theme(legend.position = "none", 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(), 
          axis.title.y = element_text(size = rel(1), angle = 0), 
          axis.text.y = element_text(size = rel(1)), 
          plot.margin = plot.margin ) 
  return(p)
}

## extract the max value of the y axis
extract_max<- function(p){
  ymax<- max(ggplot_build(p)$layout$panel_scales_y[[1]]$range$range)
  return(ceiling(ymax))
}


## main function
StackedVlnPlot<- function(obj, features,
                          pt.size = 0, 
                          plot.margin = unit(c(-0.75, 0, -0.75, 0), "cm"),
                          ...) {
  
  plot_list<- purrr::map(features, function(x) modify_vlnplot(obj = obj,feature = x, ...))
  
  # Add back x-axis title to bottom plot. patchwork is going to support this?
  plot_list[[length(plot_list)]]<- plot_list[[length(plot_list)]] +
    theme(axis.text.x=element_text(), axis.ticks.x = element_line())
  
  # change the y-axis tick to only max value 
  ymaxs<- purrr::map_dbl(plot_list, extract_max)
  plot_list<- purrr::map2(plot_list, ymaxs, function(x,y) x + 
                            scale_y_continuous(breaks = c(y)) + 
                            expand_limits(y = y))

  p<- patchwork::wrap_plots(plotlist = plot_list, ncol = 1)
  return(p)
  
#### Now able to use: StackedVlnPlot() to create a stacked violin plot that is easy to read with multipe features
}
```


## Fractioned Cluster Stacked Violin Plots - signature genes

```{r}
# Set features to test - Key signature genes are used here
EC_sig_genes <- c(
  "Cldn5",
  "Itm2a",
  "Ly6c1",
  "Ly6a",
  "Flt1",
  "Pltp",
  "Slc2a1",
  "Slco1a4"
)

MG_sig_genes <- c(
  "Cx3cr1",
  "Hexb",
  "Tmem119",
  "P2ry12",
  "Csf1r",
  "C1qa",
  "Ctss",
  "Ccl3"
)

ChPlex_sig_genes <- c(
  "Folr1",
  "Enpp2",
  "Kcnj13",
  "Ttr",
  "Prlr"
)
```

```{r}
# Create stacked violin plots to quickly visualize expression
# Run in Console (Preview Works Well, then save)
# Any feature that is not detected will cause an error. Features not found must be commented out and then rerun.

StackedVlnPlot(obj = seurat_subset_labeled_full, features = EC_sig_genes)
StackedVlnPlot(obj = seurat_subset_labeled_full, features = MG_sig_genes)
StackedVlnPlot(obj = seurat_subset_labeled_full, features = ChPlex_sig_genes)
```

### Feature plots for individual genes of interest per cluster

```{r}
# Feature plots can be done as well

# expression data is stored in the integrated assay of the object
DefaultAssay(seurat_subset_labeled_full) <- "integrated"

# set your features to test
plot.features <- c(
 "Gad2", # GABAergic Neuron
  "Robo2",
  "Ermn", # Myelin-Forming Oligodendrocytes
  "Mog",
 "Hexb",
 "C1qa"
)
  
# makes a feature plot showing expression overlaid on the UMAP. Order = T: brings highlighted cells to front. keep.scale = "all": keeps the scale the same for all feature plots made
All_Feature_Plot <- FeaturePlot(seurat_subset_labeled_full, features = plot.features, cols = c("lightgrey", "red"), order = T, label = F, pt.size = 0.05, keep.scale = "all", min.cutoff = 0, max.cutoff = 5) &
  NoLegend() &
  NoAxes()
All_Feature_Plot

# VlnPlot(seurat_subset_labeled_full, features = plot.features) + NoLegend() # can also make violin plots
```

```{r}
# Feature plots can be done as well (EC, MG, or Chplex Specific)

# expression data is stored in the integrated assay of the object
DefaultAssay(seurat_MG_subset) <- "integrated"

# these are MG specific genes
plot.features <- c("Tmem119",
                   "Hexb",
                   "Cx3cr1",
                   "C1qa",
                   "Ctss",
                   "Ccl3",
                   "P2ry12",
                   "Trem2"
                   )

FeaturePlot(seurat_MG_subset, features = plot.features, cols = c("lightgrey", "red")) # makes a feature plot showing expression overlaid on the UMAP

VlnPlot(seurat_MG_subset, features = plot.features) + NoLegend() # can also make individual violin plots to look at expression
```

```{r}
# expression data is stored in the integrated assay of the object
DefaultAssay(seurat_MG_subset) <- "integrated"

# these are more MG specific genes
MG_test <- c(
  "P2ry12",
  "Tmem119",
  "Cx3cr1",
  "Selplg",
  "Serinc3",
  "Marcks",
  "Glul",
  "Txnip",
  "Hexb",
  "Sparc",
  "Csf1r",
  "C1qa",
  "C1qb",
  "C1qc",
  "Cst3",
  "Ctss",
  "Olfml3",
  "P2ry13",
  "Tgfbr1",
  "Ctsb",
  "Apoe",
  "Lyz2",
  "Tyrobp",
  "Gnas",
  "Ftn1",
  "B2m",
  "Cstb",
  "Timp2",
  "H2-D1",
  "Trem2",
  "Axl",
  "Ctsl",
  "Lpl",
  "Cd9",
  "Csf1",
  "Ccl6",
  "Cd63",
  "Itgax",
  "Ank",
  "Pine2",
  "Cadm1",
  "Ctsz",
  "Ctsa",
  "Cd68",
  "Cd52",
  "Gusb",
  "Hif1a",
  "Ifitm3",
  "Isg15",
  "Irf7",
  "Stat1",
  "Mki67"
)

DotPlot(seurat_MG_subset, features = MG_test, split) # can make a dotplot
```


## Heatmaps

```{r}
# Set features to test - these are signature genes identified for all of the clusters
all_features <- c(
  "Cldn5", # EC Clusters
  "Itm2a",
  "Ly6c1",
  "Ly6a",
  "Flt1",
  "Cx3cr1", # MG Clusters
  "Hexb",
  "Tmem119",
  "P2ry12",
  "Csf1r",
  "Folr1", # ChPlex Clusters
  "Enpp2",
  "Kcnj13",
  "Vtn", # Pericyte
  "Pdgfrb",
  "Kcnj8",
  "Atp13a5",
  "Il34",
  "Acta2", # Arterial SMC
  "Tagln",
  "Tmp1",
  "Tmp2",
  "Crip1",
  "Rgs5", # Arterial SMC / Pericyte Doublet
  "Mylk",
  "Myh11",
  "Gkn5",
  "Gria2", # Astrocyte
  "Bcan",
  "Ttyh1",
  "Nlsr2",
  "Aqp4",
  "Tmem212", # Ependymal Cell
  "Ccdc153",
  "Foxj1",
  "Rarres2",
  "Odf3b",
  "Ccl5", # T Cell
  "Ms4a4b",
  "Trbc2",
  "Cd3d",
  "Cd3g",
  "Pf4", # Border Associated Macrophage
  "Mrc1",
  "F13a1",
  "Ms4a7",
  "Apoe",
  "Dcn", # Fibroblast
  "Spp1",
  "Lum",
  "Col1a1",
  "Mfap4",
  "CCr2", # Ccr2+ Monocyte / Macrophage
  "H2-Eb1",
  "H2-Aa",
  "Cd44",
  "Cd74",
  "Aldoc", # Reactive Astrocyte
  "S100b",
  "Mt1",
  "Mt2",
  "Mt3",
  "Olig2", # OPC 1
  "Pdgfra",
  "Sox10",
  "Cspg4",
  "Npy", # NPY+ Cell
  "Cd74", # B Cell
  "Ly6d",
  "H2-Ab1",
  "Ms4a1",
  "Cd79a",
  "Olig1", # OPC 2
  "Opcml",
  "Slc6a1",
  "Grin1", # Glutamatergic Neuron
  "Slc17a7",
  "Trank1",
  "Snhg11",
  "Kcnq2",
  "Gad2", # GABAergic Neuron
  "Robo2",
  "Sox11",
  "Nrxn3",
  "Stmn2",
  "Ermn", # Myelin-Forming Oligodendrocytes
  "Mog",
  "Mal",
  "Aplp1",
  "Plp"
)

# these are genes specific to EC cells
EC_fraction_features <- c(
  "Slc7a5",
  "Slc16a1",
  "Mfsd2a",
  "Gkn3",
  "Stmn2",
  "Bmx",
  "Jun",
  "Junb",
  "Fos",
  "Icam1",
  "Vcam1",
  "Vwf",
  "Plvap",
  "Plpp1",
  "Cd24a"
)

# these are genes specific to MG cells
MG_fraction_features <- c(
  "Ccl3",
  "Ccl4",
  "Ifngr1",
  "Il1a",
  "Gadd45",
  "Nfkbia",
  "Fos",
  "Fosb",
  "Egr1"
)

## these are genes specific to ChPlex cells
# ChPlex_fraction_features <- c(
#   "Folr1",
#   "Enpp2",
#   "Kcnj13",
#   "Ttr",
#   "Pp1r1b"
# )

```

```{r}
# set the colors used for each cluster in the heatmap - these colors will be used as the color bar above the heatmap and correspond to the colors used in the colored UMAP
colors <- c("cyan",
          "magenta",
          # "deepskyblue",
          "magenta4",
          "orange4",
          # "maroon2",
          "mediumorchid3",
          # "plum",
          # "lightslateblue",
          "limegreen",
          "darkslateblue",
          "darkcyan",
          # "tan1",
          "seagreen4",
          # "sienna1",
          "palegreen3",
          "olivedrab1",
          "darkseagreen2",
          "red4",
          "darkblue",
          "firebrick3",
          "mediumspringgreen",
          "deeppink1",
          "yellowgreen",
          "grey45",
          "darkgreen",
          "red",
          "grey75",
          "gold",
          "blue",
          "orange",
          "grey30")

EC_colors <- c( "cyan",
          "deepskyblue",
          "lightslateblue",
          "darkslateblue",
          "darkcyan",
          "darkblue",
          "blue")
  
MG_colors <- c("magenta",
          "magenta4",
          "maroon2",
          "mediumorchid3",
          "plum")

```

```{r}
# Making the heatmap for all clusters

# Make Integrated the default assay (it contains all of the genetic data in scale.data)
DefaultAssay(seurat_subset_labeled_full) <- "integrated"

# Run HeatMap for all_features (remove the NoLegend function if a legend is desired) - any genes not identified will not be used
all_heatmap <- DoHeatmap(seurat_subset_labeled_full, features = all_features, label = F, group.colors = colors, lines.width = 5) + 
  NoLegend() +
  theme(axis.text.y = element_text(size = 10)) + 
  scale_fill_gradientn(colors = c("skyblue4", "skyblue3", "oldlace", "tomato", "tomato3"), na.value = "black")
all_heatmap

## Run DotPlot for all features (optional)
# DotPlot(seurat_subset_labeled_full, features = all_features) + RotatedAxis()
```

```{r}
# Making the heatmap for the EC, MG, and ChPlex clusters

# Make Integrated the default assay (it contains all of the genetic data in scale.data)
DefaultAssay(seurat_EC_subset) <- "integrated"
DefaultAssay(seurat_MG_subset) <- "integrated"
# DefaultAssay(seurat_ChPlex_subset) <- "integrated"

# Run heatmaps for EC, MG, and ChPlex fractions (remove the NoLegend function if a legend is desired)
ec_heatmap <- DoHeatmap(seurat_EC_subset, features = EC_fraction_features, label = F, group.colors = EC_colors, lines.width = 5) +
  NoLegend() +
  theme(axis.text.y = element_text(size = 10)) +
  scale_fill_gradientn(colors = c("skyblue4", "skyblue3", "oldlace", "tomato", "tomato3"), na.value = "black")
ec_heatmap

mg_heatmap <- DoHeatmap(seurat_MG_subset, features = MG_fraction_features, label = F, group.colors = MG_colors, lines.width = 5) + 
  NoLegend() + 
  theme(axis.text.y = element_text(size = 10)) +
  scale_fill_gradientn(colors = c("skyblue4", "skyblue3", "oldlace", "tomato", "tomato3"), na.value = "black")
mg_heatmap

# DoHeatmap(seurat_ChPlex_subset, features = ChPlex_fraction_features, group.by= "seurat_clusters", label = F) + NoLegend()
```


### BBB Heatmap

```{r}
# Make the heatmap for the Brain Barriers heatmap

#Subset the full seurat object to only keep cells associated with brain barriers
BBB_Seurat <- subset(seurat_subset_labeled_full, idents = c(
  "Capillary EC",
  "Arterial EC",
  "Venous EC",
  "Pericyte",
  "Arterial SMC",
  "Astrocyte",
  "Microglia",
  "Homeostatic Microglia",
  "Border Associated Macrophage",
  "Choroid Plexus Cell"), 
  invert = FALSE)

## visualize the object subset as a UMAP
# DimPlot(object = BBB_Seurat, 
#         reduction = "umap")

# these are genes specific to the brain barriers clusters used in the heatmap (some are removed to condense the heatmap)
BBB_fraction_features <- c(
  "Cldn5", # General EC Clusters
  # "Itm2a",
  "Ly6c1",
  "Slc7a5", # Capillary EC
  # "Slc16a1",
  "Mfsd2a",
  "Gkn3", # Arterial EC
  "Stmn2",
  # "Bmx", 
  # "Icam1", # Venous EC
  "Vcam1",
  "Vwf",
  "Vtn", # Pericyte
  # "Pdgfrb",
  "Kcnj8",
  "Acta2", # Arterial SMC
  "Tagln",
  # "Tpm2",
  "Gria2", # Astrocyte
  "Bcan",
  # "Aqp4",
  "Cx3cr1", # General MG Clusters (homeostatic)
  "Hexb",
  # "Tmem119",
  # "P2ry12",
  "Ccl3", # Microglia (Activated)
  "Ccl4",
  # "Il1a", 
  "Pf4", # Border Associated Macrophage
  "Mrc1",
  # "F13a1",
  "Folr1", # ChPlex Clusters
  "Enpp2"
  # "Kcnj13"
)

# set the colors used for each cluster in the heatmap - these colors will be used as the color bar above the heatmap and correspond to the colors used in the colored UMAP
BBB_colors <- c(
          "cyan",
          "magenta",
          # "deepskyblue",
          "magenta4",
          "orange4",
          # "maroon2",
          # "mediumorchid3",
          # "plum",
          # "lightslateblue",
          "limegreen",
          "darkslateblue",
          # "darkcyan",
          # "tan1",
          "seagreen4",
          # "sienna1",
          # "palegreen3",
          "olivedrab1",
          # "darkseagreen2",
          # "red4",
          "darkblue",
          "firebrick3"
          # "mediumspringgreen",
          # "deeppink1",
          # "yellowgreen",
          # "grey45",
          # "darkgreen",
          # "red",
          # "grey75",
          # "gold",
          # "blue",
          # "orange",
          # "grey30"
)

## visualize the recolored object subset as a UMAP
# DimPlot(object = BBB_Seurat, 
        # reduction = "umap", pt.size = 1, repel = T, label.box = T, cols = c(BBB_colors))

# Make Integrated the default assay (it contains all of the genetic data in scale.data)
DefaultAssay(BBB_Seurat) <- "integrated"

# Run heatmap brain barrier subset (remove the NoLegend function if a legend is desired)
bbb_heatmap <- DoHeatmap(BBB_Seurat, features = BBB_fraction_features, label = F, size = 3, group.colors = BBB_colors, lines.width = 5) + 
  NoLegend() +
  theme(axis.text.y = element_text(size = 10)) +  
  scale_fill_gradientn(colors = c("skyblue4", "skyblue3", "oldlace", "tomato", "tomato3"), na.value = "black")
bbb_heatmap
```


# Extended EC and MG Heatmaps

```{r}
# Set features to test - these are signature genes identified for all of the clusters

# these are genes specific to EC cells
# external files can also be used to input gene lists. The file must be a .csv file, have no formatting and have a column title of "Gene"
EC_fraction_features_extended <- read.csv(file = "D:/Research/NIH/Herkenham_Lab/10X_Neurovascular_scRNAseq/Gene_Lists/EC_gene_list_focused.csv")
EC_fraction_features_extended <- as.matrix(EC_fraction_features_extended) # converts to matrix
EC_fraction_features_extended <- as.character(EC_fraction_features_extended) # converts to character

# these are genes specific to MG cells
MG_fraction_features_extended <- read.csv(file = "D:/Research/NIH/Herkenham_Lab/10X_Neurovascular_scRNAseq/Gene_Lists/MG_gene_list_focused.csv")
MG_fraction_features_extended <- as.matrix(MG_fraction_features_extended)
MG_fraction_features_extended <- as.character(MG_fraction_features_extended)
```

```{r}
# set the colors used for each cluster in the heatmap - these colors will be used as the color bar above the heatmap and correspond to the colors used in the colored UMAP

EC_colors <- c( "cyan",
          "deepskyblue",
          "lightslateblue",
          "darkslateblue",
          "darkcyan",
          "darkblue",
          "blue")
  
MG_colors <- c("magenta",
          "magenta4",
          "maroon2",
          "mediumorchid3",
          "plum")
```

```{r}
# Making the heatmap for the EC, MG, and ChPlex clusters

# Make Integrated the default assay (it contains all of the genetic data in scale.data)
DefaultAssay(seurat_EC_subset) <- "integrated"
DefaultAssay(seurat_MG_subset) <- "integrated"

# Run heatmaps for EC, MG, and ChPlex fractions (remove the NoLegend function if a legend is desired)
EXTENDEDec_heatmap <- DoHeatmap(seurat_EC_subset, features = EC_fraction_features_extended, label = F, group.colors = EC_colors, lines.width = 5) +
  NoLegend() +
  theme(axis.text.y = element_text(size = 10)) +
  scale_fill_gradientn(colors = c("skyblue4", "skyblue3", "oldlace", "tomato", "tomato3"), na.value = "black")
EXTENDEDec_heatmap

EXTENDEDmg_heatmap <- DoHeatmap(seurat_MG_subset, features = MG_fraction_features_extended, label = F, group.colors = MG_colors, lines.width = 5) + 
  NoLegend() + 
  theme(axis.text.y = element_text(size = 10)) +
  scale_fill_gradientn(colors = c("skyblue4", "skyblue3", "oldlace", "tomato", "tomato3"), na.value = "black")
EXTENDEDmg_heatmap
```







